File: src/654_MaximumBinaryTree.java
class MaximumBinaryTree {
  int[] nums;
  public TreeNode constructMaximumBinaryTree(int[] nums) {
    this.nums = nums;
    return dfs(0, nums.length-1);
  }

  private TreeNode dfs(int i, int j) {
    if (i>j) return null;

    int k=i;
    for (int l=i+1; l<=j; ++i)
      if (nums[l]>nums[k])
        k=l;

    TreeNode node = new TreeNode(nums[k]);
    node.left = dfs(i, k-1);
    node.right = dfs(k+1, j);
    return node;
  }
}

File: src/654_MaximumBinaryTree.cpp
#define _LL_DELETE_RETURN_POINTER
#include <local_leetcode.hpp>

class Solution {
  private:
    typedef vector<int>::iterator iter_t;

    TreeNode* constructImpl(iter_t begin, iter_t end) {
      if (begin == end)
        return nullptr;

      iter_t iter = max_element(begin, end);

      return new TreeNode(
        *iter,
        constructImpl(begin, iter),
        constructImpl(iter+1, end)
      );
    }

  public:
    TreeNode* constructMaximumBinaryTree(vector<int>& nums) {
      return constructImpl(nums.begin(), nums.end());
    }
};


int main() {
  EXECS(Solution::constructMaximumBinaryTree);
  return 0;
}